src/embed/connection.rs

```diff
@@ -186,30 +156,62 @@ impl<'a, 'b> Context<'a, 'b> {
     }
 }
 
-async fn execute<'a, 'b, T, S, F, E>(ctx: Context<'a, 'b>, func: F, req: Request)
-where
-    T: serde::de::DeserializeOwned,
-    S: serde::Serialize,
-    E: std::fmt::Debug,
-    F: AsyncFn2<Context<'a, 'b>, T, Output = Result<S, E>>,
-{
-    let request: T = match deserialize(&req.data.as_string().unwrap()) {
-        Ok(v) => v,
-        Err(e) => return req.response.send(Err(e)).await,
-    };
+#[derive(Debug)]
+enum ExecuteError {
+    TransactionNotFound(u32),
+    TransactionRequired,
+    TransactionIsReadOnly(u32),
+}
 
-    let result = func
-        .call(ctx, request)
-        .await
-        .map(|v| serde_json::to_string(&v));
-    // When https://doc.rust-lang.org/std/result/enum.Result.html lands we can remove this.
-    let result = match result {
-        Ok(Ok(v)) => Ok(v),
-        Ok(Err(v)) => Err(to_debug(v)),
-        Err(v) => Err(to_debug(v)),
+async fn execute<'a, 'b>(ctx: Context<'a, 'b>, req: &Request) -> Result<String, String> {
+    use ExecuteError::*;
+
+    // transaction-less
+    match req.rpc.as_str() {
+        "getRoot" => return to_json(do_get_root(ctx, from_json(&req.data)?).await),
+        "openTransaction" => return to_json(do_open_transaction(ctx, from_json(&req.data)?).await),
+        "commitTransaction" => return to_json(do_commit(ctx, from_json(&req.data)?).await),
+        "closeTransaction" => return to_json(do_close(ctx, from_json(&req.data)?).await),
+        "beginSync" => return to_json(do_begin_sync(ctx, from_json(&req.data)?).await),
+        "maybeEndSync" => return to_json(do_maybe_end_sync(ctx, from_json(&req.data)?).await),
+        _ => (),
     };
 
-    req.response.send(result).await
+    // require read txn
+    let txn_req: TransactionRequest = from_json(&req.data)?;
```

Comment:
Part of the trick I realized is to parse the input data twice -- once to get its `transaction_id` and again to get the per-rpc params. This is non-ideal, but I think worth it. Otherwise, it is very difficult to find a way to share code between the transaction-requiring and non-transaction-requiring paths.

Reply:
def worth it. didn't even occur to me.